### PR TITLE
feat(bake): retry daytona snapshot create with rich diagnostics

### PR DESCRIPTION
**Pull request by dbworku**
**TOOLCHAIN RELEASE — restage the monolithic publish job into a parallel, resumable, plan-driven pipeline.**
Shipped as an 8-PR stack. The trunk merges bottom-up: the four CLI PRs (#155–#157) are pure-additive
mechanism (nothing calls them yet), and #158/#159 are the wiring that finally consumes them.

```
main
 └─ #152  ci: pin actions/* to the current latest (checkout v7, upload v7, download v8)
     └─ #153  bake: retry the daytona snapshot create + rich failure diagnostics
         └─ #154  bake: `--provider <id>` selection + a file-based report
             └─ #155  cli: `release-plan` — the release plan as an artifact
                 └─ #156  cli: `build-candidate` — build once, record the digest
                     └─ #157  cli: `release-{validate,summary,ensure-public}` via @actions/core
                         └─ #158  ci: the five composite actions the release jobs share
                             └─ #159  ci: restage toolchain-image.yml (plan → build → bake → promote)
```

↑ **This PR is #153 (2/8) — a standalone reliability fix, based on #152.**

---

**Stack 2/8** — based on #152

Daytona's snapshot create inspects the image in the registry **on a Daytona build runner**, and that
inspect intermittently fails with an opaque "internal error" — a different runner/registry blip each
time. A single attempt turns a transient blip into a failed release.

- **Retry** the create up to 5× with linear backoff (5s → 20s), sweeping any error-state snapshot
  holding the name before each retry so the next attempt isn't rejected with "already exists". A
  transient failure now self-heals.
- **Characterize** a persistent failure instead of losing it. Every failed attempt logs the structured
  SDK error via the new `describeDaytonaError` (status code, error code, response body, cause chain —
  the SDK's error object is opaque and its `.message` alone says nothing), plus *where the snapshot
  landed*. "absent" vs an `error`-state snapshot with its `errorReason` is what distinguishes a failed
  **registry inspect** from a failed **build**.

`describeDaytonaError` is pure and defensive (it never throws — it runs on an error path), so it is
unit-tested directly without the SDK. The destroyed-snapshot contract is unchanged: it still fires only
when pre-existing snapshots were deleted, now after all attempts are exhausted.

Independent of the release restructure above it; it just happens to make that release far less flaky.

**LOC**: 152 added / 18 removed.

---

## Validation

Every branch in this stack was checked out **in isolation** (on top of its own parent, with nothing
from the PRs above it) and run through the full gate set. A reviewer merges bottom-up, so each PR has
to stand on its own — this is what verifies that.

**This PR (`02`), verified standalone — all gates green:**

| gate | command | result |
|---|---|---|
| install | `bun install --frozen-lockfile` | ✅ |
| typecheck | `bun run typecheck` | ✅ |
| lint | `bun run lint` (biome, `--error-on-warnings`) | ✅ |
| spell | `mise exec -- typos` | ✅ |
| shell | `bun run lint:shell` (shellcheck) | ✅ |
| catalog drift | `bun run check:catalog-drift` | ✅ |
| workflow lint | `actionlint` + `zizmor` | ✅ |
| tests | `bun run test` (whole workspace) | ✅ **615 passing, 0 failing** |

The same matrix is green on all 8 branches, and the passing-test count climbs monotonically up the
trunk (612 → 633) — each PR adds coverage and none regresses it. The top of the stack was also diffed
against the pre-split branch to prove the decomposition moved every change and dropped nothing.

**What this PR's own tests prove:** `describeDaytonaError` extracts status/code/body/cause from the
opaque SDK error object, includes the cause chain, and — the important one — **never throws** on a
non-object error, since it runs on the error path where a second throw would mask the original failure.

---
## Summary by cubic
Makes Daytona snapshot creation resilient and easier to debug so transient registry/runner blips self-heal, and persistent failures are clearly explained.

- **New Features**
  - Retries create up to 5 times with linear backoff (5s→20s); logs attempt number, duration, and success.
  - Cleans up error-state snapshots before each retry; stops early if the name can’t be freed and explains why.
  - Adds `describeDaytonaError` and post-failure snapshot checks to surface status/code/body/cause and where the attempt landed; preserves destroyed-snapshot behavior.

- **Bug Fixes**
  - Guarded `describeDaytonaError` against non-serializable response bodies (fallback to String) so diagnostics never throw on the error path.

<sup>Written for commit 5674596fabfdc49ada9e5c23c10928892d0c038c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/starslingdev/sandbox-benchmarks/pull/153?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added Graphite MCP server configuration to the CLI tooling.

* **Bug Fixes**
  * Improved Daytona snapshot creation reliability with up to 5 retry attempts.
  * Added safety cleanup between retries to avoid “already exists” conflicts.
  * Better final error handling when recovery cleanup is interrupted.

* **Diagnostics**
  * Enhanced Daytona error reporting with status, error code, response details, and full cause chains.

* **Tests**
  * Added coverage for detailed Daytona error formatting and edge cases.

### CODE REVIEW

**Comment by dbworku** [comment]
* **#159** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/159?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#158** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/158?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#157** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/157?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#156** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/156?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#155** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#154** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#153** <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/153?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/starslingdev/sandbox-benchmarks/153?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>merge-ready</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/starslingdev?ref=merge-queue-instructions-in-stack-comment&prId=153).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: apps/cli/src/lib/bake/daytona.ts [hunk 1]
```diff
@@ -153,23 +224,60 @@ export async function bakeDaytonaSnapshot(name: string, image: string, log: Log)
...
+		// what the destroyed-message reports — these are our own failed remnants, not a pre-existing one.)
+		if (attempt > 1) {
+			try {
+				await deleteExistingSnapshots(daytona, name, log);
+			} catch (err) {
+				log(
+					`    attempt ${attempt}: could not pre-clean ${name} — ${err instanceof Error ? err.message : String(err)}`,
+				);
+			}
+		}
```
[the remark this answers is not in the document]
Comment (dbworku) [reply apps/cli/src/lib/bake/daytona.ts:411]: Good catch — fixed, and you're right about the reasoning. `deleteExistingSnapshots` already absorbs transient blips itself (it polls to a 3-minute deadline before throwing), so a throw from it is a *persistent* failure to free the name, not something a retry can help. Continuing just burned another 3-minute pre-clean per attempt and then buried the real blocker under a generic "already exists". The loop now `break`s on pre-clean failure, sets `lastErr` to the pre-clean error so the *actual* cause is what surfaces, and logs that the name is still held.

File: apps/cli/src/lib/bake/daytona.ts [hunk 2]
```diff
@@ -153,23 +229,68 @@ export async function bakeDaytonaSnapshot(name: string, image: string, log: Log)
...
+				// for a 3-minute deadline before throwing — so a throw here is a PERSISTENT failure to free
+				// the name, not a blip. The name is still held, every remaining create would just fail
+				// "already exists" after burning another 3-minute pre-clean, and the real blocker (the
+				// undeletable snapshot) would be buried under a generic create error. Fail fast on it.
+				log(
+					`!!! attempt ${attempt}: could not free ${name} — ${err instanceof Error ? err.message : String(err)}`,
+				);
+				log("    the name is still held, so further create attempts cannot succeed; giving up.");
+				lastErr = err;
+				break;
```
[the remark this answers is not in the document]
Comment (dbworku) [reply apps/cli/src/lib/bake/daytona.ts:409]: Fixed — good catch. The pre-clean error now goes into a separate `cleanupErr`, and `lastErr` stays the create failure. The final message reports **both** when a cleanup stopped the retries (`<create error>; retries stopped because <name> could not be freed: <cleanup error>`), and the `cause` chain still points at the create error — so nothing is overwritten or misattributed.

**Review by cosmic-emerald — APPROVED** [review]

File: apps/cli/src/lib/bake/daytona.ts [hunk 3]
```diff
@@ -128,6 +128,49 @@ async function deleteExistingSnapshots(daytona: Daytona, name: string, log: Log)
...
+	if (e.code !== undefined) parts.push(`code=${String(e.code)}`);
+	if (e.response?.data !== undefined) {
+		let body: string;
+		try {
+			body =
+				typeof e.response.data === "string" ? e.response.data : JSON.stringify(e.response.data);
+		} catch {
+			body = String(e.response.data);
+		}
+		parts.push(`response=${body.slice(0, 500)}`);
```
[the remark this answers is not in the document]
Comment (dbworku) [reply apps/cli/src/lib/bake/daytona.ts:165]: Fixed in d97b62f. `describeDaytonaError` now falls back to `String(e.response.data)` when `JSON.stringify` returns `undefined` (function/symbol top-level body) instead of calling `.slice` on undefined, so the formatter can't throw on the error path. Added a regression test with a non-serializable (`() => …`) response body.

File: apps/cli/src/lib/bake/daytona.ts [hunk 4]
```diff
@@ -289,29 +365,88 @@ export async function bakeDaytonaSnapshot(name: string, image: string, log: Log)
...
+				// the name, not a blip. The name is still held, every remaining create would just fail
+				// "already exists" after burning another 3-minute pre-clean, and the real blocker (the
+				// undeletable snapshot) would be buried under a generic create error. Fail fast on it.
+				log(
+					`!!! attempt ${attempt}: could not free ${name} — ${err instanceof Error ? err.message : String(err)}`,
+				);
+				log("    the name is still held, so further create attempts cannot succeed; giving up.");
+				cleanupErr = err;
+				break;
+			}
```
[the remark this answers is not in the document]
Comment (dbworku) [reply apps/cli/src/lib/bake/daytona.ts:410]: Intentional — this is the idempotency contract of `bakeDaytonaSnapshot`, not retry-specific behavior. The function already deletes every same-name snapshot *before* the initial create (the pre-loop `deleted` sweep), so sweeping the name again between attempts is consistent with that contract rather than a new risk. These snapshots are rebuildable release *candidates*: a create that landed server-side but lost its response is simply re-created idempotently on the next attempt (ending in a valid snapshot), and the pre-loop `deleted` count is what the destroyed-artifact reporting keys off. Gating deletion on SDK failure-states would reintroduce the exact `already exists` rejection this sweep exists to prevent. Leaving as-is by design.
